kubernetes-deployment/massdriver.yaml

```diff
@@ -18,61 +18,98 @@ params:
   properties:
     namespace:
       title: Kubernetes Namespace
-      type: string
+      description: Application will be deployed into this namespace. If the namespace doesn't exist, it will be created.
+      $ref: https://raw.githubusercontent.com/massdriver-cloud/artifact-definitions/main/definitions/types/k8s-resource-name.json
       default: default
     image:
       title: Container Image
       type: object
+      required:
+      - repository
+      - tag
       properties:
         repository:
+          $md.immutable: true
           title: Repository
           # Container repositories can be created or imported into massdriver at: https://app.massdriver.cloud/container-repositories
           # You'll generally want to hard code your container repository in your bundle code, but it can be helpful to show it
           # in your schema as well.
           # If you need to allow an end-user developer to select the container repo, you can use our drop down widget documented here: https://docs.massdriver.cloud/bundles/custom-widgets-and-fields
-          description: "Container repository to use for this application."
+          description: Container repository to use for this application.
           type: string
-          enum:
-            - nginxdemos/hello
         tag:
-          title: Tag
-          description:
-          type: string
-          minLength: 1
+          title: Image Tag
+          $ref: https://raw.githubusercontent.com/massdriver-cloud/artifact-definitions/main/definitions/types/docker-image-tag.json
           default: latest
     resourceRequests:
       type: object
+      required:
+      - cpu
+      - memory
       properties:
         cpu:
           title: CPU
           description: The minimum CPU resources required for this application.
-          type: string
-          # TODO: use custom rjsf widget here
+          type: number
+          minimum: 0.001
         memory:
           title: Memory
           description: The minimum Memory resources required for this application.
-          type: string
-          # TODO: use custom rjsf widget here
+          type: number
+          minimum: .001
     port:
       title: Port
```

Comment:
It'd be great to make Port conditionally required if `ingress` is `true` but I don't think you can do `dependentRequired` across separate objects.

Reply:
I believe you still need to know the port the container is listening on